microprofile/connectors/kafka/src/main/java/io/helidon/microprofile/connectors/kafka/KafkaConnectorFactory.java

```diff
@@ -0,0 +1,130 @@
+
+    /**
+     * Microprofile messaging Kafka connector name.
+     */
+    static final String CONNECTOR_NAME = "helidon-kafka";
+    private static final String POOL_SIZE = "kafka.connector.pool.size";
+    private static final Logger LOGGER = Logger.getLogger(KafkaConnectorFactory.class.getName());
+    private final ScheduledExecutorService scheduler;
+    private final Queue<BasicKafkaConsumer<Object, Object>> consumers = new LinkedList<>();
+    private final Queue<BasicKafkaProducer<Object, Object>> producers = new LinkedList<>();
+
+    @Inject
+    KafkaConnectorFactory(Config config) {
+        scheduler = Executors.newScheduledThreadPool(config.get(POOL_SIZE)
+                .asInt().asOptional().orElseGet(() -> 10));
+    }
+
+    /**
+     * Called when container is terminated.
+     *
+     * @param event termination event
+     */
+    public void terminate(@Observes @BeforeDestroyed(ApplicationScoped.class) Object event) {
+        LOGGER.info("Terminating KafkaConnectorFactory...");
+        // Stops the scheduler first to make sure no new task will be triggered meanwhile consumers are closing
+        scheduler.shutdown();
+        BasicKafkaConsumer<Object, Object> consumer;
+        while ((consumer = consumers.poll()) != null) {
+            consumer.close();
+        }
+        BasicKafkaProducer<Object, Object> producer;
+        while ((producer = producers.poll()) != null) {
+            producer.close();
+        }
+        LOGGER.info("KafkaConnectorFactory terminated successfuly");
+    }
+
+    public Collection<BasicKafkaConsumer<Object, Object>> getConsumers() {
+        return consumers;
+    }
+
+    @Override
+    public PublisherBuilder<? extends Message<?>> getPublisherBuilder(org.eclipse.microprofile.config.Config config) {
+        Config helidonConfig = (Config) config;
+        BasicKafkaConsumer<Object, Object> basicKafkaConsumer = new BasicKafkaConsumer<>(helidonConfig, scheduler);
+        consumers.add(basicKafkaConsumer);
+        return basicKafkaConsumer.createPushPublisherBuilder();
+    }
+
+    @Override
+    public SubscriberBuilder<? extends Message<?>, Void> getSubscriberBuilder(org.eclipse.microprofile.config.Config config) {
+        Config helidonConfig = (Config) config;
+        BasicKafkaProducer<Object, Object> basicKafkaProducer = new BasicKafkaProducer<>(helidonConfig);
+        producers.add(basicKafkaProducer);
+        return ReactiveStreams.fromSubscriber(new Subscriber<Message<?>>() {
+
+            @Override
+            public void onSubscribe(Subscription s) {
+                s.request(Long.MAX_VALUE);
+            }
+
+            @Override
+            public void onNext(Message<?> message) {
+                LOGGER.fine("On next received " + message.getPayload());
+                basicKafkaProducer.produce(message.getPayload());
+                message.ack();
+            }
+
+            @Override
+            public void onError(Throwable t) {
```

Comment:
You should close the producer in `onError` as well, as `onComplete` may never be called.

Reply:
Done